crates/iceberg/src/transform/temporal.rs

```diff
@@ -0,0 +1,343 @@
+        Ok(Arc::<Int32Array>::new(
+            binary(
+                month_array.as_any().downcast_ref::<Int32Array>().unwrap(),
+                year_array.as_any().downcast_ref::<Int32Array>().unwrap(),
+                // Compute month from 1970-01-01, so minus 1 here.
+                |a, b| a + b - 1,
+            )
+            .unwrap(),
+        ))
+    }
+}
+
+/// Extract a date or timestamp day, as days from 1970-01-01
+pub struct Day;
+
+impl TransformFunction for Day {
+    fn transform(&self, input: ArrayRef) -> Result<ArrayRef> {
+        let res: Int32Array = match input.data_type() {
+            DataType::Timestamp(datatypes::TimeUnit::Microsecond, _) => input
+                .as_any()
+                .downcast_ref::<TimestampMicrosecondArray>()
+                .unwrap()
+                .unary(|v| -> i32 { (v as f64 / 1000.0 / 1000.0 * DAY_PER_SECOND) as i32 }),
+            DataType::Date32 => {
+                input
+                    .as_any()
+                    .downcast_ref::<Date32Array>()
+                    .unwrap()
+                    .unary(|v| -> i32 {
+                        datatypes::Date32Type::to_naive_date(v).num_days_from_ce()
+                            - DAY_SINCE_UNIX_EPOCH
+                    })
+            }
+            _ => {
+                return Err(Error::new(
+                    ErrorKind::Unexpected,
+                    format!(
+                        "Should not call internally for unsupport data type {:?}",
+                        input.data_type()
+                    ),
+                ))
+            }
+        };
+        Ok(Arc::new(res))
+    }
+}
+
+/// Extract a timestamp hour, as hours from 1970-01-01 00:00:00
+pub struct Hour;
+
+impl TransformFunction for Hour {
+    fn transform(&self, input: ArrayRef) -> Result<ArrayRef> {
+        let res: Int32Array = match input.data_type() {
+            DataType::Timestamp(datatypes::TimeUnit::Microsecond, _) => input
+                .as_any()
+                .downcast_ref::<TimestampMicrosecondArray>()
+                .unwrap()
+                .unary(|v| -> i32 { (v as f64 * HOUR_PER_SECOND / 1000.0 / 1000.0) as i32 }),
+            _ => {
+                return Err(Error::new(
+                    ErrorKind::Unexpected,
+                    format!(
+                        "Should not call internally for unsupport data type {:?}",
+                        input.data_type()
+                    ),
+                ))
+            }
+        };
+        Ok(Arc::new(res))
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use arrow::array::{ArrayRef, Date32Array, Int32Array, TimestampMicrosecondArray};
+    use chrono::NaiveDate;
+    use std::sync::Arc;
+
+    use crate::transform::TransformFunction;
+
+    #[test]
+    fn test_transform_years() {
+        let year = super::Year;
+        let ori_date = vec![
+            NaiveDate::from_ymd_opt(1970, 1, 1).unwrap(),
```

Comment:
Can we also add `timestamp`, `timestamptz`?

Reply:
Yes. And I think here NaiveDateTime(timestamp) and DataTime(timestamptz) have the same effect. I add the NaiveDateTime to cover more case. 🤔 Please let me know if something I miss.